src/components/tl/ucp/reduce/reduce.c

```diff
@@ -13,21 +13,27 @@ ucc_status_t ucc_tl_ucp_reduce_knomial_finalize(ucc_coll_task_t *task);
 
 ucc_status_t ucc_tl_ucp_reduce_init(ucc_tl_ucp_task_t *task)
 {
-    task->super.post     = ucc_tl_ucp_reduce_knomial_start;
-    task->super.progress = ucc_tl_ucp_reduce_knomial_progress;
-    task->super.finalize = ucc_tl_ucp_reduce_knomial_finalize;
     ucc_coll_args_t   *args      = &task->super.args;
     ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
     ucc_rank_t         myrank    = team->rank;
     ucc_rank_t         team_size = team->size;
-    ucc_rank_t         root      = (uint32_t)args->root;
+    ucc_rank_t         root      = args->root;
     ucc_rank_t         vrank     = (myrank - root + team_size) % team_size;
-    ucc_memory_type_t  mtype     = args->src.info.mem_type;
     ucc_status_t       status    = UCC_OK;
-    size_t             data_size =
-        args->src.info.count * ucc_dt_size(args->src.info.datatype);
+    ucc_memory_type_t  mtype;
+    size_t             data_size;
     int                isleaf;
 
+    if (root == myrank) {
```

Comment:
i think its more readable if you'll do
ucc_coll_buffer_info_t info = (root == myrank) ? args->dst.info ? args->src.info 

and then just do the rest once